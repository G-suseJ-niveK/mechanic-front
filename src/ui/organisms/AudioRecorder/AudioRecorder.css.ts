import { makeStyles, Theme } from '@material-ui/core/styles';

export default makeStyles((theme: Theme) => ({
  contentButton: {
    [theme.breakpoints.up('sm')]: {
      padding: '0rem 9rem !important'
    }
  },
  button: {
    width: '100%',
    borderRadius: '50px'
  },
  buttonDefault: {
    background: '#828282',
    color: 'white',
    '&:hover': {
      backgroundColor: '#828282a8 !important'
    }
  },
  recording: {
    width: '100%',
    background: '#9E2A2A',
    borderRadius: '50px',
    '&:hover': {
      background: '#bb4242'
    }
  },
  microphone: {
    display: 'flex',
    padding: '30px',
    borderRadius: '30px',
    alignItems: 'center',
    justifyContent: 'center'
  },
  microphoneDisabled: {
    color: '#DADADA',
    background: '#828282'
  },
  microphoneEnabled: {
    color: '#DADADA',
    background: '#9E2A2A',
    animation: `$flicker 2000ms ${theme.transitions.easing.easeInOut}`,
    animationIterationCount: 'infinite'
  },
  audio: {
    width: '100%',
    [theme.breakpoints.up('sm')]: {
      padding: '0.25rem 5rem !important'
    },
    '&:focus': {
      outline: 'none'
    }
  },
  audioOutline: {
    width: '100%',
    '&:focus': {
      outline: 'none'
    }
  },
  '@-moz-keyframes flicker': {
    '0%': { opacity: ' 1.0' },
    '50%': { opacity: '0.5' },
    '00%': { opacity: '1.0' }
  },
  '@-webkit-keyframes flicker': {
    '0%': { opacity: ' 1.0' },
    '50%': { opacity: '0.5' },
    '00%': { opacity: '1.0' }
  },
  '@keyframes flicker': {
    '0%': { opacity: ' 1.0' },
    '50%': { opacity: '0.5' },
    '00%': { opacity: '1.0' }
  }
}));
