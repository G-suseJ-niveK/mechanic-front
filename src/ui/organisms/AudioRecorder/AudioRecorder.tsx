import React, { useState, useCallback, useEffect } from 'react';
import { Button, Icon, Typography, Grid, Box, Divider } from '@material-ui/core';
import useStyles from './AudioRecorder.css';
import clsx from 'clsx';

let mediaRecorder: any = null;
let audioFragments: any = [];

type AudioRecorderProps = {
  onSuccess?(audio: Blob): void;
  onError?(error: string): void;
  disabled?: boolean;
  inputSource?: any;
  time?: number; // value on seconds
  textTime?: string;
};

const AudioRecorder: React.FC<AudioRecorderProps> = (props: AudioRecorderProps) => {
  const classes = useStyles();
  const { onSuccess, onError, textTime, time, inputSource, disabled } = props;
  const [isRecordingActive, setIsRecordingActive] = useState<boolean>(false);
  const [audioSRC, setAudioSRC] = useState<string | null>(null);
  const [idInterval, setIdInterval] = useState<any>(null);
  const [currentSeconds, setCurrentSeconds] = useState<number>(0);
  const [elapsedTime, setElapsedTime] = useState<string>('00:00:00');

  const handleStopRecording = useCallback(() => {
    try {
      mediaRecorder.stop();
      mediaRecorder = null;
      clearInterval(idInterval);
    } catch (error) {}
  }, [idInterval]);

  const startCounting = useCallback(() => {
    const tiempoSegundos = Date.now();
    const auxIdInterval = setInterval(() => {
      setCurrentSeconds((Date.now() - tiempoSegundos) / 1000);
    }, 1000);
    setIdInterval(auxIdInterval);
  }, []);

  const verifyMicrophonePermission = async () => {
    const result = await navigator.permissions
      .query({ name: 'microphone' })
      .then((res: PermissionStatus) => {
        if (res.state === 'granted') {
          return true;
        }
        if (res.state === 'prompt') {
          return true;
        }
        if (res.state === 'denied') {
          return false;
        }
      })
      .catch(() => {
        return false;
      });
    return result;
  };

  const handleStartRecording = useCallback(() => {
    if (!verifyMicrophonePermission()) {
      onError && onError('Problemas al grabar, verifique que tenga un microfono disponible');
      return;
    }

    navigator.mediaDevices
      .getUserMedia({
        audio: true,
        video: false
      })
      .then((stream: any) => {
        audioFragments = [];
        // eslint-disable-next-line
        // @ts-ignore
        mediaRecorder = new MediaRecorder(stream);
        mediaRecorder.start();
        setIsRecordingActive(true);
        startCounting();
        mediaRecorder.addEventListener('dataavailable', (evento: any) => {
          // Add audio
          audioFragments.push(evento.data);
        });

        mediaRecorder.addEventListener('stop', () => {
          // Stop el stream
          stream.getTracks().forEach((track: any) => track.stop());

          // Convert audio
          const blobAudio = new Blob(audioFragments || [], { type: 'audio/ogg' });
          setAudioSRC(URL.createObjectURL(blobAudio));
          onSuccess && onSuccess(blobAudio);
        });
      })
      .catch(() => {
        onError && onError('Problemas al grabar, verifique que tenga un microfono disponible');
      });
  }, [onError, onSuccess, startCounting]);

  const handleReplayRecording = useCallback(() => {
    setElapsedTime('00:00:00');
    setCurrentSeconds(0);
    setAudioSRC(null);
    audioFragments = [];
    setIsRecordingActive(false);
  }, []);

  const convertToTime = useCallback(
    (segundos: number) => {
      let seconds = segundos;

      let hh: string = '';
      let mm: string = '';
      let ss: string = '';
      const h: number = Math.floor(seconds / 60 / 60);
      seconds -= h * 60 * 60;
      const m: number = Math.floor(seconds / 60);
      seconds -= m * 60;
      const s: number = Math.floor(seconds);
      hh = h < 10 ? '0' + h : String(h);
      mm = m < 10 ? '0' + m : String(m);
      ss = s < 10 ? '0' + s : String(s);
      setElapsedTime(`${hh}:${mm}:${ss}`);
      if (time && time + 1 <= segundos) {
        handleStopRecording();
      }
    },
    [handleStopRecording, time]
  );

  useEffect(() => {
    convertToTime(currentSeconds);
  }, [currentSeconds, convertToTime]);

  useEffect(() => {
    if (audioSRC === null) {
      inputSource && setAudioSRC(inputSource);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <>
      <Grid container spacing={1}>
        {!audioSRC && (
          <Grid
            item
            xs={12}
            sm={12}
            md={12}
            lg={12}
            xl={12}
            style={{ flexDirection: 'column', display: 'flex', alignItems: 'center' }}>
            <Icon
              fontSize="large"
              className={clsx(
                classes.microphone,
                !isRecordingActive && classes.microphoneDisabled,
                isRecordingActive && classes.microphoneEnabled
              )}>
              mic
            </Icon>

            <Typography>{elapsedTime}</Typography>
          </Grid>
        )}
        {!isRecordingActive && !audioSRC && (
          <Grid item xs={12} sm={12} md={12} lg={12} xl={12} className={classes.contentButton}>
            <Button
              disabled={disabled || false}
              className={clsx(classes.button, classes.buttonDefault)}
              variant="contained"
              size="large"
              startIcon={<Icon>fiber_manual_record_icon</Icon>}
              onClick={handleStartRecording}>
              <Typography>Comenzar a Grabar</Typography>
            </Button>
          </Grid>
        )}
        {isRecordingActive && !audioSRC && (
          <Grid item xs={12} sm={12} md={12} lg={12} xl={12} className={classes.contentButton}>
            <Button
              disabled={disabled || false}
              className={clsx(classes.button)}
              style={{ color: 'white' }}
              variant="contained"
              color="secondary"
              size="large"
              startIcon={<Icon>stop</Icon>}
              onClick={handleStopRecording}>
              Detener grabaci√≥n
            </Button>
          </Grid>
        )}
        {!audioSRC && textTime && (
          <Grid item xs={12} sm={12} md={12} lg={12} xl={12}>
            <Divider style={{ border: '1px solid #96C262' }} />
            <Typography align="center" style={{ color: '#828282' }}>
              {textTime}
            </Typography>
          </Grid>
        )}
        {audioSRC && (
          <Grid item xs={12} sm={12} md={12} lg={12} xl={12}>
            <audio controls className={clsx(classes.audio)}>
              <source src={audioSRC} type="audio/ogg" />
              El navegador no soporta audio.
            </audio>
            <Box width="100%" className={classes.contentButton}>
              <Button
                disabled={disabled || false}
                className={clsx(classes.button)}
                variant="contained"
                color="primary"
                size="large"
                startIcon={<Icon>replay</Icon>}
                onClick={handleReplayRecording}>
                Volver a grabar
              </Button>
            </Box>
          </Grid>
        )}
      </Grid>
    </>
  );
};

export default React.memo(AudioRecorder);
