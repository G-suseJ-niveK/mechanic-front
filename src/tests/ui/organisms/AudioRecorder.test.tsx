import React from 'react';
import { render, screen } from '@testing-library/react';
import AudioRecorder from '~organisms/AudioRecorder/AudioRecorder';

describe('Render AudioRecorder', () => {
  // const children: React.ReactNode = <div>children</div>;

  test('should render', () => {
    const { container } = render(<AudioRecorder />);
    expect(container).toMatchSnapshot();
  });

  test('should disabled', () => {
    render(<AudioRecorder disabled={true} />);
    expect(screen.getByRole('button')).toBeDisabled();
  });

  test('should render text', () => {
    render(<AudioRecorder textTime="test text" />);
    expect(screen.getByText(/test text/i)).not.toBeNull();
  });

  // test('should recording', () => {
  //   let audioResult: Blob | null = null;
  //   const mockOnSuccess = (audio: Blob) => {
  //     audioResult = audio;
  //   };

  //   render(<AudioRecorder time={2} onSuccess={mockOnSuccess} />);
  //   fireEvent.click(screen.getByRole('button'));
  // });
});
