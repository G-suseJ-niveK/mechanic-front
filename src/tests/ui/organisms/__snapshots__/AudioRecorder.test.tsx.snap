// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render AudioRecorder should render 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-12 MuiGrid-grid-md-12 MuiGrid-grid-lg-12 MuiGrid-grid-xl-12"
      style="flex-direction: column; display: flex; align-items: center;"
    >
      <span
        aria-hidden="true"
        class="material-icons MuiIcon-root makeStyles-microphone-5 makeStyles-microphoneDisabled-6 MuiIcon-fontSizeLarge"
      >
        mic
      </span>
      <p
        class="MuiTypography-root MuiTypography-body1"
      >
        00:00:00
      </p>
    </div>
    <div
      class="MuiGrid-root makeStyles-contentButton-1 MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-12 MuiGrid-grid-md-12 MuiGrid-grid-lg-12 MuiGrid-grid-xl-12"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button-2 makeStyles-buttonDefault-3 MuiButton-containedSizeLarge MuiButton-sizeLarge"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          <span
            class="MuiButton-startIcon MuiButton-iconSizeLarge"
          >
            <span
              aria-hidden="true"
              class="material-icons MuiIcon-root"
            >
              fiber_manual_record_icon
            </span>
          </span>
          <p
            class="MuiTypography-root MuiTypography-body1"
          >
            Comenzar a Grabar
          </p>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
